bigint/big_integer.cpp

```diff
@@ -1,242 +1,550 @@
 #include "big_integer.h"
-
-#include <cstring>
+#include <cmath>
+#include <algorithm>
 #include <stdexcept>
 
-big_integer::big_integer()
-{
-    mpz_init(mpz);
+void trim(big_integer &a, size_t s) {
+    while (a.digits.size() < s) {
+        a.digits.push_back(a.further);
+    }
+}
+
+void shrink(big_integer &a) {
+    while (a.digits.size() > 1 && a.digits.back() == a.further) {
+        a.digits.pop_back();
+    }
+}
+
+uint32_t big_integer::operator[](size_t index) const {
+    return (index < digits.size() ? digits[index] : further);
+}
+
+static std::function<uint32_t(uint32_t, uint32_t, uint32_t&)> adc =
+        [](uint32_t a, uint32_t b, uint32_t &d) {
+    uint32_t c = a + b + d;
+    d = (c < a || (d == 1 && c <= a));
+    return c;
+};
+
+static std::function<uint32_t(uint32_t, uint32_t, uint32_t&)> sbb =
+        [](uint32_t a, uint32_t b, uint32_t &d) {
+    uint32_t c = a + ~b + d;
+    d = (c < a || (d == 1 && c <= a));
+    return c;
+};
+
+static std::function<uint32_t(uint32_t, uint32_t)> and_ =
+        [](uint32_t a, uint32_t b) { return a & b; };
+
+static std::function<uint32_t(uint32_t, uint32_t)> or_ =
+        [](uint32_t a, uint32_t b) { return a | b; };
+
+static std::function<uint32_t(uint32_t, uint32_t)> xor_ =
+        [](uint32_t a, uint32_t b) { return a ^ b; };
+
+big_integer &evaluate(big_integer &a, const big_integer &b, uint32_t d,
+    std::function<uint32_t(uint32_t, uint32_t, uint32_t&)> &eval) {
+    trim(a, b.digits.size());
+    a.digits.push_back(a.further);
+    for (size_t i = 0; i < a.digits.size(); ++i) {
+        a.digits[i] = eval(a.digits[i], b[i], d);
+    }
+    a.further = (a.digits.back() & 2 ? ~0 : 0);
+    shrink(a);
+
+    return a;
+}
+
+big_integer &bitwise(big_integer &a, const big_integer &b,
+    std::function<uint32_t(uint32_t, uint32_t)> &eval) {
+    trim(a, b.digits.size());
+    for (size_t i = 0; i < a.digits.size(); ++i) {
+        a.digits[i] = eval(a.digits[i], b[i]);
+    }
+    a.further = eval(a.further, b.further);
+    shrink(a);
+
+    return a;
+}
+
+int unsigned_cmp(const big_integer &a, const big_integer &b) {
+    for (size_t i = std::max(a.digits.size(), b.digits.size()) - 1;; --i) {
```

Comment:
[design] Можно сделать вот так:

```suggestion
    for (size_t i = std::max(a.digits.size(), b.digits.size()); i-- > 0;) {
```

И `if` не нужен.